tests/SpecTests/PrimaryStepDownSpecTest.php

```diff
@@ -0,0 +1,251 @@
+        $this->assertCount(1, $events);
+        $this->assertSame('getMore', $events[0]['started']->getCommandName());
+
+        // Verify that no new connections have been created
+        $this->assertSame($totalConnectionsCreated, $this->getTotalConnectionsCreated());
+
+        // Wait to allow primary election to complete and prevent subsequent test failures
+        $this->waitForMasterReelection();
+    }
+
+    public function testNotMasterKeepsConnectionPool()
+    {
+        $runOn = [(object) ['minServerVersion' => '4.1.11', 'topology' => [self::TOPOLOGY_REPLICASET]]];
+        $this->checkServerRequirements($runOn);
+
+        // Set a fail point
+        $this->configureFailPoint([
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => [
+                'failCommands' => ['insert'],
+                'errorCode' => 10107,
+            ],
+        ]);
+
+        $totalConnectionsCreated = $this->getTotalConnectionsCreated();
+
+        // Execute an insert into the test collection of a {test: 1} document.
+        try {
+            $this->insertDocuments(1);
+        } catch (BulkWriteException $e) {
+            // Verify that the insert failed with an operation failure with 10107 code.
+            $this->assertSame(10107, $e->getCode());
+        }
+
+        // Execute an insert into the test collection of a {test: 1} document and verify that it succeeds.
+        $result = $this->insertDocuments(1);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        // Verify that the connection pool has not been cleared
+        $this->assertSame($totalConnectionsCreated, $this->getTotalConnectionsCreated());
+    }
+
+    public function testNotMasterResetConnectionPool()
+    {
+        $runOn = [(object) ['maxServerVersion' => '4.0.999', 'topology' => [self::TOPOLOGY_REPLICASET]]];
+        $this->checkServerRequirements($runOn);
+
+        // Set a fail point
+        $this->configureFailPoint([
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => [
+                'failCommands' => ['insert'],
+                'errorCode' => 10107,
+            ],
+        ]);
+
+        $totalConnectionsCreated = $this->getTotalConnectionsCreated();
+
+        // Execute an insert into the test collection of a {test: 1} document.
+        try {
+            $this->insertDocuments(1);
+        } catch (BulkWriteException $e) {
+            // Verify that the insert failed with an operation failure with 10107 code.
+            $this->assertSame(10107, $e->getCode());
+        }
+
+        // Verify that the connection pool has been cleared
+        $this->assertSame($totalConnectionsCreated + 1, $this->getTotalConnectionsCreated());
+    }
+
+    public function testShutdownResetConnectionPool()
+    {
+        $runOn = [(object) ['minServerVersion' => '4.0.0', 'topology' => [self::TOPOLOGY_REPLICASET]]];
```

Comment:
https://github.com/mongodb/specifications/tree/master/source/connections-survive-step-down/tests#shutdown-in-progress-reset-connection-pool states that this and "Interrupted at shutdown - Reset Connection Pool" should be run on all server versions. Other tests only report a version number, which is a bit vague.

I assume that the getMore test requires a replica set since it invokes a `replSetStepDown` command, but should these other tests run on standalones and sharded clusters? For a sharded cluster, the test will need to use a single mongos in its URI to ensure the fail point is properly applied/used.

Reply:
> https://github.com/mongodb/specifications/tree/master/source/connections-survive-step-down/tests#shutdown-in-progress-reset-connection-pool states that this and "Interrupted at shutdown - Reset Connection Pool" should be run on all server versions. Other tests only report a version number, which is a bit vague.

I was confused about that as well. The [original spec commit](https://github.com/mongodb/specifications/commit/cb1c5fea75a3d44f1368b073ea91aa00ba3d338c#diff-8caf393de24930695605b96dec90e5baR113) specified "version 4.0 or lower" for `testNotMasterResetConnectionPool` and "all supported server versions" for the two shutdown tests. This was later changed in https://github.com/mongodb/specifications/commit/1e043d971e9f7793e5e5d4e68985a17313df643a to exclude server versions below 4.0 as `failCommand` was only added in 4.0. That's why this there is a single version requirement instead of a range.

> I assume that the getMore test requires a replica set since it invokes a replSetStepDown command, but should these other tests run on standalones and sharded clusters? For a sharded cluster, the test will need to use a single mongos in its URI to ensure the fail point is properly applied/used.

I'd limit the `getMore` and both `not master` tests to replica sets as the `not master` error shouldn't ever appear on non-replicase topologies. The other tests definitely should run on standalones and sharded clusters as well.